File: gomsic_core/parsers/gomsoftware_cfg.py
"""Parser for gomsoftware.cfg (T.O.M. configuration format).

Sources:
- gomsic/local-config/<version>/gomsoftware.cfg
- gomsic/config/<version>/ (other config files)

The T.O.M. config format is an XML-like structure with nested sections.
"""

from __future__ import annotations

import logging
import re
from typing import Any, Optional

from ..debug_log.trace import ParserTraceContext
from ..extractor import ArchiveLayout
from ..models import GomSoftwareConfig
from .base import BaseParser

logger = logging.getLogger(__name__)


class GomSoftwareCfgParser(BaseParser):
    name = "gomsoftware_cfg"

    def parse(self, layout: ArchiveLayout, ctx: ParserTraceContext) -> Optional[GomSoftwareConfig]:
        if layout.gomsic_dir is None:
            ctx.skip("gomsic directory not found")
            return None

        # Look for gomsoftware.cfg in local-config/<version>/ directories
        cfg_files = self.find_files(layout.gomsic_dir, "local-config/*/gomsoftware.cfg")
        if not cfg_files:
            cfg_files = self.find_files(layout.gomsic_dir, "config/*/gomsoftware.cfg")
        if not cfg_files:
            ctx.skip("gomsoftware.cfg not found")
            return None

        # Use the most recent version (last in sorted order)
        cfg_path = cfg_files[-1]
        ctx.file_searched(str(cfg_path))
        ctx.file_found(str(cfg_path))

        text = self.read_text_file(cfg_path)
        if text is None:
            ctx.fail(f"Could not read {cfg_path}")
            return None

        ctx.file_parsed(str(cfg_path))

        config = GomSoftwareConfig(raw_text=text)
        config.sections = self._parse_tom_config(text)

        ctx.note(f"Parsed {len(config.sections)} top-level sections from {cfg_path.name}")
        return config

    def _parse_tom_config(self, text: str) -> dict[str, dict[str, Any]]:
        """Parse T.O.M. config format into nested dictionaries.

        The format is roughly:
            section_name {
                key = value
                subsection {
                    key = value
                }
            }
        """
        sections: dict[str, dict[str, Any]] = {}
        current_section: Optional[str] = None
        current_data: dict[str, Any] = {}
        brace_depth = 0

        for line in text.splitlines():
            stripped = line.strip()
            if not stripped or stripped.startswith("#") or stripped.startswith("//"):
                continue

            # Opening brace with section name
            open_match = re.match(r"^(\S+)\s*\{", stripped)
            if open_match and brace_depth == 0:
                current_section = open_match.group(1)
                current_data = {}
                brace_depth = 1
                continue

            if "{" in stripped:
                brace_depth += stripped.count("{")
            if "}" in stripped:
                brace_depth -= stripped.count("}")
                if brace_depth <= 0:
                    if current_section:
                        sections[current_section] = current_data
                    current_section = None
                    current_data = {}
                    brace_depth = 0
                continue

            # Key = value within a section
            if current_section and brace_depth == 1:
                kv_match = re.match(r"^(\S+)\s*=\s*(.*)$", stripped)
                if kv_match:
                    key = kv_match.group(1)
                    value = kv_match.group(2).strip().strip('"').strip("'")
                    current_data[key] = value

        return sections

File: pwa/src/worker/parsers/gomsoftware-cfg.ts
// ============================================================
// parsers/gomsoftware-cfg.ts — Port of Python GomSoftwareCfgParser
// Parses T.O.M. brace-delimited config files (gomsoftware.cfg)
// ============================================================

import { BaseParser, ParserContext } from './base'
import type { GomSoftwareConfig } from '../models'
import { gomsicConfigDir } from '../extractor'

export class GomSoftwareCfgParser extends BaseParser<GomSoftwareConfig> {
  readonly name = 'gomsoftware_cfg'

  parse(ctx: ParserContext): GomSoftwareConfig | null {
    const { layout, gomsicDir } = ctx
    if (!gomsicDir) return null

    // Search for gomsoftware.cfg in config directories
    const candidates: string[] = []

    // Try gomsicConfigDir (checks local-config, all-config, config)
    const configDir = gomsicConfigDir(layout)
    if (configDir) {
      const found = this.findFiles(layout, configDir, '*/gomsoftware.cfg')
      candidates.push(...found)
    }

    // Also explicitly try local-config and config if gomsicConfigDir didn't find them
    for (const sub of ['local-config', 'config']) {
      const dir = `${gomsicDir}/${sub}`
      const found = this.findFiles(layout, dir, '*/gomsoftware.cfg')
      for (const f of found) {
        if (!candidates.includes(f)) candidates.push(f)
      }
    }

    if (candidates.length === 0) return null

    // Use the first found file (prefer local-config over config)
    const vpath = candidates[0]
    const text = this.getText(layout, vpath)
    if (!text) return null

    const sections = this._parseTomConfig(text)

    return {
      sections,
      raw_text: text,
    }
  }

  /**
   * Parse T.O.M. brace-delimited config format.
   * Only extracts key=value pairs at depth == 1 (direct children of top-level sections).
   *
   * Example:
   *   section_name {
   *       key = value
   *       subsection {
   *           key = value  <- ignored (depth 2)
   *       }
   *   }
   */
  private _parseTomConfig(text: string): Record<string, Record<string, unknown>> {
    const result: Record<string, Record<string, unknown>> = {}
    let depth = 0
    let currentSection = ''
    const sectionStack: string[] = []

    for (const rawLine of text.split(/\r?\n/)) {
      const line = rawLine.trim()
      if (!line || line.startsWith('#') || line.startsWith('//')) continue

      // Check for opening brace (possibly "name {" on same line)
      if (line.includes('{')) {
        const nameMatch = line.match(/^(\S+)\s*\{/)
        const sectionName = nameMatch ? nameMatch[1] : ''

        if (depth === 0) {
          currentSection = sectionName
          if (currentSection && !(currentSection in result)) {
            result[currentSection] = {}
          }
          sectionStack.push(currentSection)
        } else {
          sectionStack.push(sectionName)
        }
        depth++
        continue
      }

      // Closing brace
      if (line === '}') {
        if (depth > 0) {
          sectionStack.pop()
          depth--
          if (depth === 0) {
            currentSection = ''
          }
        }
        continue
      }

      // Key=value pair — only extract at depth 1 (direct section children)
      if (depth === 1 && currentSection && line.includes('=')) {
        const eqIdx = line.indexOf('=')
        const key = line.slice(0, eqIdx).trim()
        const value = line.slice(eqIdx + 1).trim()
        if (key) {
          result[currentSection][key] = this._parseValue(value)
        }
      }
    }

    return result
  }

  /**
   * Attempt to coerce values to appropriate types.
   * Strings that look like numbers become numbers; "true"/"false" become booleans.
   */
  private _parseValue(raw: string): unknown {
    // Strip surrounding quotes
    if (
      (raw.startsWith('"') && raw.endsWith('"')) ||
      (raw.startsWith("'") && raw.endsWith("'"))
    ) {
      return raw.slice(1, -1)
    }
    if (raw.toLowerCase() === 'true') return true
    if (raw.toLowerCase() === 'false') return false
    const num = Number(raw)
    if (!isNaN(num) && raw.length > 0) return num
    return raw
  }
}
